### ISSUE TITLE
Invisible data causes conflict when deleting a page

### ISSUE THREAD

**Issue by danrot** [sulu/sulu#4378]
| Q | A
| --- | ---
| Bug? | yes
| New Feature? | no
| Sulu Version | 22b3b41fab8a3edcf34d489d8a0925b06efe7a39
| Browser Version | ---

#### Actual Behavior

When a page is referenced on another page, it cannot be deleted, because a 409 response is returned. This is what is expected, but when the template of the page or the type of a block is changed, the references in PHPCR still exist, and the linked page is still recognized as linked, although the content manager does not see that anywhere.

#### Expected Behavior

If the link is not visible to the content manager, resp. the property containing the link is not part of the current template, it should not be recognized as linked by the system, because this behavior is really strange for the content manager.

#### Steps to Reproduce

1. Have a template with a block property with two different types, whereby one contains an `internal_links` field, e.g.
```xml
        <block name="blocks" default-type="editor" minOccurs="0">
            <meta>
                <title lang="de">Inhalte</title>
                <title lang="en">Content</title>
            </meta>
            <types>
                <type name="editor">
                    <meta>
                        <title lang="de">Editor</title>
                        <title lang="en">Editor</title>
                    </meta>
                    <properties>
                        <property name="article" type="text_editor">
                            <meta>
                                <title lang="de">Artikel</title>
                                <title lang="en">Article</title>
                            </meta>
                        </property>
                    </properties>
                </type>
                <type name="links">
                    <meta>
                        <title lang="de">Links</title>
                        <title lang="en">Links</title>
                    </meta>
                    <properties>
                        <property name="link" type="internal_links">
                            <meta>
                                <title lang="de">Artikel</title>
                                <title lang="en">Article</title>
                            </meta>
                        </property>
                    </properties>
                </type>
            </types>
        </block>
```
2. Create a page "Page A"
3. Create a page "Page B" using the template with the block step 1, and add a block with an `internal_links` field type and link to "Page A" and save the page
4. Change the type of the block to show the editor instead, so that the link to "Page B" is not visible to the content manager anymore and save again
5. Try to delete "Page A", and see that the request fails

#### Possible Solutions

The problem wouldn't be that bad, if the behavior on the frontend would be correctly implemented (showing a second dialog explaining where the content is linked and allow to force delete anyway).

However, even if this is implemented it is strange, if a content manager tries to find the link in the content by themselves, because it does not seem to be there. Therefore we should probably try to recognize if the found reference is still valid in terms of being visible to the content manager.

**Comment by HadGS** [comment]
I think, that correct behaviour should be little different. All data properties in blocks, that are not part of actually saved block type, should be deleted. Currently all data, that are not overwritten by new block type, are saved again. Save first block with type A, then type B, then type C, and we have three sets of properties saved in DB...

**Comment by danrot** [comment]
That's true, but that can somehow even be considered a feature, because you easily can recover old data by just switchting the template, and it is harder for the content manager to mess up. If we delete the other data it is gone after saving, and just switching back to the template won't help anymore.

It would be the cleaner solution from a technical perspective IMO, but it seems like there are content manager out there, who are glad that it is not implemented that way :slightly_smiling_face:

**Comment by niklasnatter** [comment]
Fixed in #4809
